Allow CanvasEngine to possibly have debouncing options for smooth animations

antonioru opened the pull request:
…closes #478

# Checklist

- [x] The code has been run through pretty `yarn run pretty`
- [x] The tests pass on CircleCI
- [x] You have referenced the issue(s) or other PR(s) this fixes/relates-to
- [x] The PR Template has been filled out (see below)
- [x] Had a beer/coffee because you are awesome

## What?
Performance improvement with a debounce

## Why?
Why not?

## How?
wraps the `CanvasEngine.repaintCanvas` method within a 60ms debounce

## Feel good image:

(Add your own one below :])

![LOL](https://i.pinimg.com/originals/7f/1b/c3/7f1bc3fb2e123dd3255a85c04db22f19.jpg)

pierre-moire commented:
If we use animation a frame represents around 14ms so if you use 60ms debounce the layout will never be updated

pierre-moire commented:
10-12ms debounce is enough to avoid performance issues during multiple nodes animation

dylanvorster commented:
yeah I'm not entirely sure this is the correct thing to do.

Firstly:
* If a dev wants this, they can do this via extending the DiagramEngine and overriding the repaint.
* @pierre-moire makes a good point
* React already does this internally with the Fibre engine.
* There are better approaches to getting more performance.

antonioru commented:
@dylanvorster 
I understand and yes, you're probably right! 
What about making the debounce time a parameter and possibly switch it off? 

Otherwise I think  extending the DiagramEngine it's the way and we should get rid of this PR ;)

antonioru commented:
[ not related to the PR ]
Also @dylanvorster I have a question and I'm sorry I have to do it here but I haven't your direct contact, anyway... the question is: do we have to build the library before opening a PR?

antonioru commented:
@dylanvorster 
Ok I've updated the PR by creating 2 new options for the CanvasEngine: `repaintDebounce: boolean` and `repaintDebounceMs: number`. 
The first one allows debouncing the `repaintCanvas` method as suggested by @pierre-moire whilst the second one defines the milliseconds for the debounce to be waited (by default, the debounce is set as false).
Then I updated @pierre-moire 's animation example to use those options. 

I think this should possibly "fix" all the issues you highlighted but again, if you're not comfortable with merging this PR as you think in this particular cases the developers should extend the CanvasEngine class I'm totally fine with it :) 

cheers guys

antonioru commented:
@dylanvorster sorry to bother, any news for this PR? :)

dylanvorster commented on packages/react-canvas-core/src/CanvasEngine.ts:
```diff
@@ -19,15 +20,25 @@ export interface CanvasEngineListener extends BaseListener {
+	/**
+	 * If set to true debounce the `repaintCanvas` method
+	 */
+	repaintDebounce?: boolean
```
I would prefer this to be implicit based on the repaintDebounceMs. As in, if the value is > 0 then we actually apply it without needing the `repaintDebounce ` variable

antonioru replied:
@dylanvorster all right

dylanvorster commented on packages/react-canvas-core/src/CanvasEngine.ts:
```diff
@@ -122,6 +144,9 @@ export class CanvasEngine<
 		};
 
+		// if repaintDebounce is set to true, debounce the 'repaint' function by the milliseconds defined by repaintDebounceMs
+		const repaintFn = repaintDebounce ? debounce(repaint, repaintDebounceMs) : repaint;
```
also I would prefer not using ternary operators outside of react :) Prettier should actually remove this, did you run `yarn pretty` ?

antonioru replied:
@dylanvorster Yes I did, it doesn't remove the ternary operator. 
Anyway I'll remove it but why is that?
There's a precise reason?

antonioru commented:
Ok, I've updated the PR, thank you @dylanvorster @pierre-moire

antonioru commented:
@dylanvorster @pierre-moire any news about this PR?

pierre-moire commented:
> @dylanvorster @pierre-moire any news about this PR?

Seems great to me

dylanvorster commented on packages/diagrams-demo-gallery/demos/demo-animation/index.tsx:
```diff
@@ -28,7 +27,7 @@ class NodeDelayedPosition extends React.Component<any, any> {
 export default () => {
 	//1) setup the diagram engine
-	var engine = createEngine();
+	var engine = createEngine({ repaintDebounce: true, repaintDebounceMs: 12 });
```
can remove the boolean now :)

antonioru replied:
@dylanvorster 
done! sorry, it was a typo

antonioru replied:
@dylanvorster any news about this MR? 
It's kind of needed lately :)

pierre-moire commented:
@dylanvorster Can you merge it please?
